### DISCUSSION TITLE
[Chapter 06] Enhanced switch ?

### DISCUSSION THREAD

**Discussion by coalong** [cowaine/is-my-code-that-strange#53] [Chapter 06 (조건 분기)]
책의 예시 코드를 작성하다가 IntelliJ 에서 아래와 같은 경고 표시를 보여줬다.

Enhanced switch 란 무엇일까 찾아보니 JDK 13 부터 지원하는 향상된 switch 문이라고 한다.

Enhanced switch 문은 어떤 변화가 있었으며, 여러 교육중 switch 보다 if 문을 사용하기를 권장하는데 Enhanced switch 보다도 if 가 좋을까요?

<img width="944" alt="image" src="https://github.com/cowaine/is-my-code-that-strange/assets/71637466/4c45f1ff-fef4-4a82-885c-c258f2cc9c9f">

**Comment by firewoody237** [comment]
### if와 switch 중 어떤 것이 좋은가?
https://jerry92k.tistory.com/51?category=981716
- 위 페이지를 참고하시면, `if-else`를 바이트코드로 컴파일하면 조건에 따라 최대 O(N)의 결과
- `switch`를 바이트코드를 바꾸면, 
    -  `case`가 연속적인 경우 -> `tableswitch`를 사용하여 `case`가 연속적일 경우 효율적이게 컴파일
    -  `case`가 연속적이지 않은 경우 -> `lookupswitch`로 `jump table`을 활용하여 탐색이 빠름
- 다만 다음 페이지를 보면, 옛날부터 그냥 보기 쉬운것을 쓰는게 제일 좋다 라는 결론 인 것 같아요. [페이지](https://kldp.org/node/62262) 
- 어떤 얘기로는 조건이 3개 이하면 if else, 4개 이상이면 switch가 성능적으로 더 좋다고 하네요.
- 근데 전 이런걸로 성능 신경 안쓸거같아요. 요새 컴퓨팅 성능이 좋은걸..ㅎㅎ

**Comment by corock** [comment]
전 그래도 협업자 간 사이드 이펙트 방지를 위해 `if` 문을 사용할 것 같습니다. (`else`, `switch` 사용 안 함)
우아한테크코스 부트캠프에서는 아예 `switch` 사용을 금지하는 가이드를 내리기도 합니다.

_enhanced switch_ 는 처음 보는데 이미 `switch` 를 보면 몸이 거부해서...  근데 쓸 기회가 있다면 한번 써봐야겠네요 ㅎㅎ
